src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp

```diff
@@ -0,0 +1,103 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "DataStructures/Variables.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+class DataVector;
+template <size_t Dim>
+class Index;
+namespace PUP {
+class er;
+}  // namespace PUP
+// IWYU pragma: no_forward_declare Variables
+/// \endcond
+
+/// \ingroup NumericalAlgorithmsGroup
+/// Interpolates a `Variables` onto an arbitrary set of points.
+template <size_t Dim>
+class IrregularInterpolant {
+ public:
+  IrregularInterpolant(
+      const Index<Dim>& mesh,
+      const tnsr::I<DataVector, Dim, Frame::Logical>& points) noexcept;
+  IrregularInterpolant() = default;
+
+  // clang-tidy: no runtime references
+  /// Serialization for Charm++
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  //@{
+  /// Performs the interpolation on a `Variables` with grid points corresponding
+  /// to the `Index<Dim>` specified in the constructor.
+  /// The result is a `Variables` whose internal `DataVector` goes over the
+  /// list of points that were specified in the constructor.
+  /// \note for the void function, `result` will be resized to the proper size.
+  template <typename TagsList>
+  void interpolate(gsl::not_null<Variables<TagsList>*> result,
+                   const Variables<TagsList>& u) const noexcept;
+  template <typename TagsList>
+  Variables<TagsList> interpolate(const Variables<TagsList>& u) const noexcept;
+  //@}
+
+ private:
+  // clang-tidy: redundant declaration.
+  template <size_t Dimm>
+  friend bool operator==(const IrregularInterpolant<Dimm>& /*lhs*/,  // NOLINT
+                         const IrregularInterpolant<Dimm>& /*rhs*/) noexcept;
+  Matrix interpolation_matrix_;
+};
+
+template <size_t Dim>
+template <typename TagsList>
+void IrregularInterpolant<Dim>::interpolate(
+    gsl::not_null<Variables<TagsList>*> result,
+    const Variables<TagsList>& u) const noexcept {
```

Comment:
So we have 3 types of files: hpp, cpp, and tpp.
- hpp files typically house declarations and class definitions. Function template definitions typically also go in hpp files, such as in this case. The reason is that we need to instantiate function templates, which means we need to know all the possible types that the template parameters will have in our code, otherwise we get linking errors. Knowing all the possible types is quite easy in some situations, such as the pointwise functions where it is always a double or DataVector. For Variables<Tags> it is more difficult, since we have no idea what the tags might be (not quite true, but the problem is much harder than DataVector and double). In this case it makes sense to have the compiler automatically instantiate the templates for us. However, we pay for this with increased compilation time.
- cpp files house non-function template definitions (regular functions), as well as function templates where the explicit instantiations are easy (such as templates like Dim, or DataVector/double cases). Basically, if it is quite easy to enumerate all the template parameter values and you do not want the functions inlined, putting function templates in cpp files is preferred because it reduces compile time.
- tpp files are the most confusing because they are very similar to hpp files, and will often get used as such. tpp files are designed to keep the function template definitions, but be able to include the file like a header file. The advantage is that we can include the hpp file in many places without having to pay for extra parsing and instantiations, but we must make sure that an instantiation occurs somewhere so the linker will find it. tpp files are currently used for the derivatives because we know what we are differentiating for each system, but want to have the instantiations be done in for each system without the derivative files needing to know about the system themselves. This means we include the tpp file in, say, `Equations.cpp` and generate the explicit instantiations for that system there.

Reply:
Thanks, @nilsdeppe. This is very helpful.

Reply:
@nilsdeppe is there a place in the documentation that you can put your above comment about usage of hpp/cpp/tpp files?  This is something that new developers should read.

Reply:
I'm certain there should be, I'm not sure what it is. We need something in the `Dev Guide` that covers all this sort of stuff. Maybe something like a `Core Principles` section? Thoughts on a name?